src/daemon/common.c

```diff
@@ -360,18 +360,20 @@ int strjoin (char *buffer, size_t buffer_size,
 			ptr += sep_len;
 			avail -= sep_len;
 		}
+		if (avail == 0)
+			return (-1);
 
 		field_len = strlen (fields[i]);
 		if (avail < field_len)
-			return (-1);
+			field_len = avail;
 
 		memcpy (ptr, fields[i], field_len);
 		ptr += field_len;
 		avail -= field_len;
 	}
 
 	assert (buffer[buffer_size - 1] == 0);
-	return ((int) strlen (buffer));
+	return ((int) (buffer_size - avail - 1));
```

Comment:
I think we might want, `((buffer_size - 1) - avail)` so it is clear we are subtracting the ending value of avail from its beginning value.